### ISSUE TITLE
Add support for discovery of selective refresh partials

### ISSUE THREAD

**Issue by westonruter** [Automattic/customize-direct-manipulation#23]
As noted in [Selective Refresh in the Customizer](https://make.wordpress.org/core/2016/02/16/selective-refresh-in-the-customizer/):

> ## Linking Partials to Controls
> 
> Since each partial is associated with at least one setting, shift-clicking on the placement for a partial will attempt to focus on the related control in the pane. This was already implemented for widgets since 3.9 and recently for nav menu items as well, but now the functionality is available to partials generally.

Selective refresh partials generally have a `selector` associated with them and this is how the `placements` for that partial are found. Since a partial is associated with one or more settings and a control is also associated with one or more settings, a partial can be linked to a control by means of the settings they share. This is implemented in core today whereby any partial placement will get the “Shift-click to edit” tooltip when hovering over it: https://github.com/xwp/wordpress-develop/blob/4.6.0/src/wp-includes/js/customize-selective-refresh.js#L85-L99

And when shift-clicking it will then call that `focusControl` method which sends a `focus-control-for-setting` message to the customizer pane to then focus on the control associated with that setting:
- https://github.com/xwp/wordpress-develop/blob/4.6.0/src/wp-includes/js/customize-selective-refresh.js#L171-L184
- https://github.com/xwp/wordpress-develop/blob/4.6.0/src/wp-admin/js/customize-controls.js#L4190-L4203

All of this to say, all partials could make use of the discoverability that the DM plugin makes available here. Via the `wp.customize.selectiveRefresh.Partial` class, there is a common interface for linking elements in the preview with controls in the pane, and this common interface could be extended with the DM functionality.

**Comment by sirbrillig** [comment]
This is really interesting! I haven't had a chance to fully grok partials yet, but I'll do so now. Thanks!

**Comment by sirbrillig** [comment]
I've finally had some time to look into this. Very cool! Yes, partial selectors should make it much easier to identify many items on the page.

Please excuse this rambling train of thought as I explore the next steps.

So what I need to figure out now is how to find that selector for a partial via javascript. It looks like maybe the array `wp.customize.selectiveRefresh.data.partials` inside the preview frame is supposed to contain partials, from which maybe I could get their selectors, but in all the wp.com themes I tested that array is always empty. It could be that they don't support partial refresh?

Also, since the positioning of partials already probably runs some JS, the way CDM works right now may be duplicating a lot of effort. Let me start with a different tack:

When a partial's content is rendered for the first time, or when that partial is finished updating its content, I want to run some javascript that has access to two pieces of data: the partial selector, and the ability to send `focus-setting-for-partial` to the server (possibly via the `showControl()` method of the `Partial` prototype).

If my module was in core, it looks like I may be able to do what I want by simply putting my code in the `ready()` method of `Partial`, as you've done with shift-clicks. Without editing core files, however, I'll need another way to hook into that event. I think I could do that maybe by waiting on the `Deferred` on `Partial.ready`, but I'd still have to be able to access a list of the partials to be able to attach that listener. Once again I'll need to figure out a way to get a list of partials. I'll take a look at the theme code.

**Comment by sirbrillig** [comment]
Update: turns out I just wasn't looking at the right themes. I found support in `twentythirteen`, but not `twentyfifteen` or `twentysixteen`. Experimentation commencing.

**Comment by westonruter** [comment]
@sirbrillig the list of partials are available in the preview via `wp.customize.selectiveRefresh.partial`. You can iterate over all of them via:

``` js
wp.customize.selectiveRefresh.partial.each( function( partial ) {
    // ...
} );
```

And you can listen for newly-added partials (such as when Jetpack infinite scroll happens) via:

``` js
wp.customize.selectiveRefresh.partial.bind( 'add', function( addedPartial ) {
    // ...
} );
```

And yes, you can wait for the partial to be initialized via:

``` js
partial.deferred.ready.done( function() {
    // ...
} )
```

(Aside: I think I just noticed a bug that partials that get added for new elements added to the page may not get their `ready` deferreds resolved.)

Inside of the deferred, you can then get access to all of the `placements` for that partial . Placements include the actual DOM element. So you'd be able to do something like:

``` js
_.each( _.pluck( partial.placements(), 'container' ), function( container ) {
    // ... attach the CDM UI element
} );
```
